### ISSUE TITLE
no auto-completion in spark-shell

### ISSUE THREAD

**Issue by bfblog** [bitnami/bitnami-docker-spark#85]
### Name and Version

bitnami/spark:latest

### What steps will reproduce the bug?

1. docker run -ti bitnami/spark ./bin/spark-shell

### What is the expected behavior?

I expected auto-completion capabiltities in spark-shell, e.g. tab-key to show possible completions.

### What do you see instead?

The tab-key creates a whitespace and does not auto-completion. Also cursor key do not work.

### Additional information

I created a [custom image](https://github.com/bfblog/spark-tools/blob/work/Dockerfile) and here the auto-completion works pretty find. 

This is my custom container:
```
shell> docker run -ti bfblog/spark ./bin/spark-shell
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/05/30 07:19:23 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://9f45e8298d3b:4040
Spark context available as 'sc' (master = local[*], app id = local-1653895165204).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.2.1
      /_/

Using Scala version 2.12.15 (OpenJDK 64-Bit Server VM, Java 1.8.0_322)
Type in expressions to have them evaluated.
Type :help for more information.
```
And this is the bitnami container. Here auto-completion in spark-shell does not work:
```
shell> docker run -ti bitnami/spark ./bin/spark-shell
 07:20:35.99
 07:20:35.99 Welcome to the Bitnami spark container
 07:20:35.99 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-spark
 07:20:35.99 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-spark/issues
 07:20:35.99

Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/05/30 07:20:47 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://6d9d59c843fc:4040
Spark context available as 'sc' (master = local[*], app id = local-1653895248843).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.2.1
      /_/

Using Scala version 2.12.15 (OpenJDK 64-Bit Server VM, Java 1.8.0_332)
Type in expressions to have them evaluated.
Type :help for more information.
```

**Comment by jotamartos** [comment]
Hi @bfblog ,

Thank you for taking the time to report this issue. I made the autocompletion work by installing the `readline` package. Can you run these commands and confirm the solution work for you?

```
cd /opt/bitnami/spark
. venv/bin/activate
install_packages build-essential libncurses5-dev
curl -LJO "https://raw.githubusercontent.com/certifi/python-certifi/master/certifi/cacert.pem"
mv cacert.pem ./venv/lib/python3.8/site-packages/pip/_vendor/certifi/
pip install readline
./bin/spark-shell
```

Thanks

**Comment by bfblog** [comment]
Hi @jotamartos,

I am doing the same. Starting the container and after spark is started, pressing tab and cursor keys. As you see here, tab creates some whitespaces the cursor keys create some escape-sequences. This output happens with bitnami/spark images. Doing the same with my custom image (bfblog/spark) works:
```
shell> docker run -ti bitnami/spark ./bin/spark-shell
 16:36:56.91
 16:36:56.91 Welcome to the Bitnami spark container
 16:36:56.91 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-spark
 16:36:56.91 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-spark/issues
 16:36:56.91

Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/06/01 16:37:02 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://167913561ad9:4040
Spark context available as 'sc' (master = local[*], app id = local-1654101423329).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.2.1
      /_/

Using Scala version 2.12.15 (OpenJDK 64-Bit Server VM, Java 1.8.0_332)
Type in expressions to have them evaluated.
Type :help for more information.

scala> :        ^[[B^[[A^[[C^[[D
```

And now the same with bfblog/spark:
```
 docker run -ti bfblog/spark ./bin/spark-shell
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/06/01 16:43:15 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://915ab478e5db:4040
Spark context available as 'sc' (master = local[*], app id = local-1654101797413).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.2.1
      /_/

Using Scala version 2.12.15 (OpenJDK 64-Bit Server VM, Java 1.8.0_332)
Type in expressions to have them evaluated.
Type :help for more information.

scala> :
completions   help      h?        implicits   line   paste   quit     require   save   settings   type   warnings
edit          history   imports   javap       load   power   replay   reset     sh     silent     kind

scala> :
```

It seems to be an issue with bitnami/spark and permissions. At default, the container runs as user 1001 and now I change it to root. Here completions works:
```
docker run -ti --user root bitnami/spark ./bin/spark-shell
 16:44:32.98
 16:44:32.98 Welcome to the Bitnami spark container
 16:44:32.98 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-spark
 16:44:32.98 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-spark/issues
 16:44:32.99

Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/06/01 16:44:38 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://22fa78a78cfc:4040
Spark context available as 'sc' (master = local[*], app id = local-1654101879316).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.2.1
      /_/

Using Scala version 2.12.15 (OpenJDK 64-Bit Server VM, Java 1.8.0_332)
Type in expressions to have them evaluated.
Type :help for more information.

scala> :
completions   help      h?        implicits   line   paste   quit     require   save   settings   type   warnings
edit          history   imports   javap       load   power   replay   reset     sh     silent     kind
```

I am active with the bitnami/spark image in a different slack channel:
```
shell> ./bin/spark-shell --packages io.delta:delta-core_2.12:1.2.1
 16:48:20.93
 16:48:20.93 Welcome to the Bitnami spark container
 16:48:20.93 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-spark
 16:48:20.93 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-spark/issues
 16:48:20.94

I have no name!@28026f499e8d:/opt/bitnami/spark$ ./bin/spark-shell --packages io.delta:delta-core_2.12:1.2.1
:: loading settings :: url = jar:file:/opt/bitnami/spark/jars/ivy-2.5.0.jar!/org/apache/ivy/core/settings/ivysettings.xml
Ivy Default Cache set to: /opt/bitnami/spark/.ivy2/cache
The jars for the packages stored in: /opt/bitnami/spark/.ivy2/jars
io.delta#delta-core_2.12 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-aeeb9c6a-bf4d-49b7-88f5-937b5f2c2f1c;1.0
        confs: [default]
Exception in thread "main" java.io.FileNotFoundException: /opt/bitnami/spark/.ivy2/cache/resolved-org.apache.spark-spark-submit-parent-aeeb9c6a-bf4d-49b7-88f5-937b5f2c2f1c-1.0.xml (No such file or directory)
        at java.io.FileOutputStream.open0(Native Method)
        at java.io.FileOutputStream.open(FileOutputStream.java:270)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
        at org.apache.ivy.plugins.parser.xml.XmlModuleDescriptorWriter.write(XmlModuleDescriptorWriter.java:71)
        at org.apache.ivy.plugins.parser.xml.XmlModuleDescriptorWriter.write(XmlModuleDescriptorWriter.java:63)
        at org.apache.ivy.core.module.descriptor.DefaultModuleDescriptor.toIvyFile(DefaultModuleDescriptor.java:553)
        at org.apache.ivy.core.cache.DefaultResolutionCacheManager.saveResolvedModuleDescriptor(DefaultResolutionCacheManager.java:183)
        at org.apache.ivy.core.resolve.ResolveEngine.resolve(ResolveEngine.java:259)
        at org.apache.ivy.Ivy.resolve(Ivy.java:522)
        at org.apache.spark.deploy.SparkSubmitUtils$.resolveMavenCoordinates(SparkSubmit.scala:1445)
        at org.apache.spark.util.DependencyUtils$.resolveMavenDependencies(DependencyUtils.scala:185)
        at org.apache.spark.deploy.SparkSubmit.prepareSubmitEnvironment(SparkSubmit.scala:308)
        at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:898)
        at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
        at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
        at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
        at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1043)
        at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1052)
        at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
I have no name!@28026f499e8d:/opt/bitnami/spark$ ls -alF
total 184
drwxr-xr-x 1 root root  4096 May 19 18:01 ./
drwxrwxr-x 1 root root  4096 May 19 17:59 ../
-rw-r--r-- 1 root root     0 Mar  1 14:47 .buildcomplete
-rw-r--r-- 1 root root 22878 Mar  1 14:47 LICENSE
-rw-r--r-- 1 root root 57677 Mar  1 14:47 NOTICE
drwxr-xr-x 3 root root  4096 Mar  1 15:13 R/
-rw-r--r-- 1 root root  4512 Mar  1 14:47 README.md
-rw-r--r-- 1 root root   167 Mar  1 14:47 RELEASE
drwxr-xr-x 2 root root  4096 Mar  1 15:13 bin/
drwxrwxr-x 1 root root  4096 Jun  1 16:48 conf/
drwxr-xr-x 5 root root  4096 Mar  1 15:13 data/
drwxr-xr-x 4 root root  4096 Mar  1 15:13 examples/
drwxrwxr-x 1 root root 16384 Mar  1 15:13 jars/
drwxr-xr-x 4 root root  4096 Mar  1 15:13 kubernetes/
drwxr-xr-x 2 root root  4096 Mar  1 15:13 licenses/
drwxrwxr-x 2 root root  4096 May 19 18:01 logs/
drwxr-xr-x 8 root root  4096 Mar  1 15:13 python/
drwxr-xr-x 2 root root  4096 Mar  1 15:13 sbin/
drwxrwxr-x 1 root root  4096 Jun  1 16:48 tmp/
drwxr-xr-x 4 root root  4096 Mar  1 15:13 venv/
drwxrwxr-x 2 root root  4096 May 19 18:01 work/
drwxr-xr-x 2 root root  4096 Mar  1 15:13 yarn/
```
Here, the --package option tries to download additional artifacts and fails, because container runs as user 1001 but /opt/bitnami/spark is writable by root only. Let the container run with --user root also solves this issue:
```
shell> docker run -ti --user root bitnami/spark bash
 16:52:03.47
 16:52:03.47 Welcome to the Bitnami spark container
 16:52:03.47 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-spark
 16:52:03.48 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-spark/issues
 16:52:03.48

root@8cd4c37d792d:/opt/bitnami/spark# ./bin/spark-shell --packages io.delta:delta-core_2.12:1.2.1
:: loading settings :: url = jar:file:/opt/bitnami/spark/jars/ivy-2.5.0.jar!/org/apache/ivy/core/settings/ivysettings.xml
Ivy Default Cache set to: /root/.ivy2/cache
The jars for the packages stored in: /root/.ivy2/jars
io.delta#delta-core_2.12 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-5ad0be03-7b7a-4dde-8823-3bff0dc0bd95;1.0
        confs: [default]
        found io.delta#delta-core_2.12;1.2.1 in central
        found io.delta#delta-storage;1.2.1 in central
        found org.antlr#antlr4-runtime;4.8 in central
        found org.codehaus.jackson#jackson-core-asl;1.9.13 in central
downloading https://repo1.maven.org/maven2/io/delta/delta-core_2.12/1.2.1/delta-core_2.12-1.2.1.jar ...
        [SUCCESSFUL ] io.delta#delta-core_2.12;1.2.1!delta-core_2.12.jar (311ms)
downloading https://repo1.maven.org/maven2/io/delta/delta-storage/1.2.1/delta-storage-1.2.1.jar ...
        [SUCCESSFUL ] io.delta#delta-storage;1.2.1!delta-storage.jar (19ms)
downloading https://repo1.maven.org/maven2/org/antlr/antlr4-runtime/4.8/antlr4-runtime-4.8.jar ...
        [SUCCESSFUL ] org.antlr#antlr4-runtime;4.8!antlr4-runtime.jar (52ms)
downloading https://repo1.maven.org/maven2/org/codehaus/jackson/jackson-core-asl/1.9.13/jackson-core-asl-1.9.13.jar ...
        [SUCCESSFUL ] org.codehaus.jackson#jackson-core-asl;1.9.13!jackson-core-asl.jar (43ms)
:: resolution report :: resolve 1883ms :: artifacts dl 430ms
        :: modules in use:
        io.delta#delta-core_2.12;1.2.1 from central in [default]
        io.delta#delta-storage;1.2.1 from central in [default]
        org.antlr#antlr4-runtime;4.8 from central in [default]
        org.codehaus.jackson#jackson-core-asl;1.9.13 from central in [default]
        ---------------------------------------------------------------------
        |                  |            modules            ||   artifacts   |
        |       conf       | number| search|dwnlded|evicted|| number|dwnlded|
        ---------------------------------------------------------------------
        |      default     |   4   |   4   |   4   |   0   ||   4   |   4   |
        ---------------------------------------------------------------------
:: retrieving :: org.apache.spark#spark-submit-parent-5ad0be03-7b7a-4dde-8823-3bff0dc0bd95
        confs: [default]
        4 artifacts copied, 0 already retrieved (3348kB/10ms)
22/06/01 16:52:31 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Spark context Web UI available at http://8cd4c37d792d:4040
Spark context available as 'sc' (master = local[*], app id = local-1654102356302).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.2.1
      /_/

Using Scala version 2.12.15 (OpenJDK 64-Bit Server VM, Java 1.8.0_332)
Type in expressions to have them evaluated.
```
I assume both problems are related missing file system permissions. What du you think?

**Comment by jotamartos** [comment]
Hi @bfblog,

Correct, this is a permissions issue. In order to solve this issue quickly, you just need to add a chmod command to the Dockerfile 

```
FROM docker.io/bitnami/minideb:bullseye
LABEL maintainer "Bitnami <containers@bitnami.com>"

ENV HOME="/" \
    OS_ARCH="amd64" \
    OS_FLAVOUR="debian-11" \
    OS_NAME="linux" \
    PATH="/opt/bitnami/python/bin:/opt/bitnami/java/bin:/opt/bitnami/spark/bin:/opt/bitnami/spark/sbin:/opt/bitnami/common/bin:$PATH"

ARG JAVA_EXTRA_SECURITY_DIR="/bitnami/java/extra-security"

COPY prebuildfs /
# Install required system packages and dependencies
RUN install_packages acl ca-certificates curl gzip libbz2-1.0 libc6 libcom-err2 libcrypt1 libffi7 libgcc-s1 libgssapi-krb5-2 libk5crypto3 libkeyutils1 libkrb5-3 libkrb5support0 liblzma5 libncursesw6 libnsl2 libreadline8 libsqlite3-0 libssl1.1 libstdc++6 libtinfo6 libtirpc3 procps tar zlib1g
RUN . /opt/bitnami/scripts/libcomponent.sh && component_unpack "python" "3.8.13-1" --checksum 79af9dcbaa89c4047d2d24b4a4c2ae17b771fe94972734379b9e50ef3dec3442
RUN . /opt/bitnami/scripts/libcomponent.sh && component_unpack "java" "1.8.333-0" --checksum 02a91d298bbe3bb5d240f635802a50b2970b19eba960c1804b7275f88a944bb3
RUN . /opt/bitnami/scripts/libcomponent.sh && component_unpack "spark" "3.2.1-0" --checksum b5d35eb7ab1e078e598eed9b5c237ccd86f91400f1462229f2e5a611f55e0268
RUN . /opt/bitnami/scripts/libcomponent.sh && component_unpack "gosu" "1.14.0-0" --checksum da4a2f759ccc57c100d795b71ab297f48b31c4dd7578d773d963bbd49c42bd7b
RUN apt-get update && apt-get upgrade -y && \
    rm -r /var/lib/apt/lists /var/cache/apt/archives
RUN chmod g+rwX /opt/bitnami
RUN chown -R 1001:1001 /opt/bitnami/spark

COPY rootfs /
RUN /opt/bitnami/scripts/spark/postunpack.sh
RUN /opt/bitnami/scripts/java/postunpack.sh
ENV APP_VERSION="3.2.1" \
    BITNAMI_APP_NAME="spark" \
    JAVA_HOME="/opt/bitnami/java" \
    LD_LIBRARY_PATH="/opt/bitnami/python/lib/:/opt/bitnami/spark/venv/lib/python3.8/site-packages/numpy.libs/:$LD_LIBRARY_PATH" \
    LIBNSS_WRAPPER_PATH="/opt/bitnami/common/lib/libnss_wrapper.so" \
    NSS_WRAPPER_GROUP="/opt/bitnami/spark/tmp/nss_group" \
    NSS_WRAPPER_PASSWD="/opt/bitnami/spark/tmp/nss_passwd" \
    PYTHONPATH="/opt/bitnami/spark/python/:$PYTHONPATH" \
    SPARK_HOME="/opt/bitnami/spark"

WORKDIR /opt/bitnami/spark
USER 1001
ENTRYPOINT [ "/opt/bitnami/scripts/spark/entrypoint.sh" ]
CMD [ "/opt/bitnami/scripts/spark/run.sh" ]
```

I'll apply this change in our pipeline so the image is built with that new layer. Thanks for reporting this.

**Comment by bfblog** [comment]
Thank you very much!

**Comment by jotamartos** [comment]
Just for the record, these changes are included in the solution since version `3.2.1-debian-11-r1`

https://github.com/bitnami/bitnami-docker-spark/commit/5e3501775f4657e5fada11dae0308296063b1300
